File: 3-matmult/pointers/matmult.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

// Valor constante para llenar la matriz
static double a = 1.0E-10;

void FillMatrix(double *matrixA, double *matrixB, int N);
void Multiply(double *matrixA, double *matrixB, double *matrixC, int N);
double AddMatrix(double *matrix, int N);
void OpenFile(int *N);
void SaveFile(struct tm *start, struct tm *end, double error, double elapsed, int N);

int main(int argc, char *argv[])
{
    int N;               // Dimension de la matriz
    double *matrixA;     // Primera matriz
    double *matrixB;     // Segunda matriz
    double *matrixC;     // Matriz resultado
    double result;       // Resultado de la suma de la matriz resultado
    double estimation;   // Estimacion del calculo
    double error;        // Error encontrado
    double elapsed;      // Tiempo que tomo ejecutarse el programa
    time_t t;            // Variable de tiempo
    struct tm *start;    // Hora de inicio
    struct tm *end;      // Hora de termino
    clock_t start_clock; // Tiempo de inicio
    clock_t stop_clock;  // Tiempo de termino

    // Inicia la variable de tiempo
    t = time(NULL);

    // Establece la hora de inicio
    start = localtime(&t);
    start_clock = clock();

    // Asigna la dimension de la matriz
    OpenFile(&N);

    // Reserva la memoria para las tres matrices
    matrixA = (double *)malloc(sizeof(double) * N * N);
    matrixB = (double *)malloc(sizeof(double) * N * N);
    matrixC = (double *)malloc(sizeof(double) * N * N);

    // Llena la matriz A y B con el valor constante
    FillMatrix(matrixA, matrixB, N);

    // Multiplica las matrices A y B guardando el valor en la matriz C
    Multiply(matrixA, matrixB, matrixC, N);

    //Calcula la suma de los valores de la matriz C
    result = AddMatrix(matrixC, N);

    // Calculo estimado con la formula a^2*N^3
    estimation = pow(N, 3) * pow(a, 2);

    // Calcula el porcentaje de error.
    error = fabs(result - estimation) / estimation * 100.0;

    // Libera la memoria de las tres matrices
    free(matrixA);
    free(matrixB);
    free(matrixC);

    // Establece la hora en que termino de calcular
    end = localtime(&t);
    stop_clock = clock();

    // Calcula el tiempo que tomo ejecutar el programa
    elapsed = (double)(stop_clock - start_clock) / CLOCKS_PER_SEC;

    // Guarda el resultado en un archivo
    SaveFile(start, end, error, elapsed, N);

    return 0;
}

// Llena las dos matrices con el valor constante
void FillMatrix(
    double *matrixA, // Primera matriz
    double *matrixB, // Segunda matriz
    int N            // Dimension de la matriz
)
{
    int i; // Indice el renglon
    int j; // Indice de la columna

    for (i = 0; i < N; i++)
        for (j = 0; j < N; j++)
        {
            matrixA[(i * N) + j] = a;
            matrixB[(i * N) + j] = a;
        }
}

// Multiplica las dos matrices y almacena el resultado en la matriz de resultado
void Multiply(
    double *matrixA, // Primera matriz
    double *matrixB, // Segunda matriz
    double *matrixC, // Matriz resultado
    int N            // Dimension de la matriz
)
{
    int i;         // Indice del renglon
    int j;         // Indice de la columna
    int k;         // Indice de la multiplicacion
    double result; // Resultado de la multiplicacion

    for (i = 0; i < N; i++)
        for (j = 0; j < N; j++)
        {
            result = 0.0;
            for (k = 0; k < N; k++)
                result += matrixA[(i * N) + k] * matrixB[(k * N) + j];
            matrixC[(i * N) + j] = result;
        }
}

// Suma todos los elementos de una matriz y regresa el resultado
double AddMatrix(
    double *matrix, // Matriz resultado
    int N           // Dimension de la matriz
)
{
    double result; // Resultado de la suma
    int i;         // Indice del renglon
    int j;         // Indice de la columna

    result = 0.0;

    for (i = 0; i < N; i++)
        for (j = 0; j < N; j++)
            result += matrix[(i * N) + j];

    return result;
}

// Abre un archivo con la dimension de la matriz
void OpenFile(
    int *N // Dimension de la matriz
)
{
    FILE *file;

    file = fopen("parameters.dat", "r");

    if (file == NULL)
        printf("No se puede abrir el archivo.\n");
    else
    {
        fscanf(file, "%d", N);
        fclose(file);
    }
}

// Crea un archivo de salida con la hora de inicio, de termino y el tiempo que tomo correr el programa
// Asi como el porcentaje de error
void SaveFile(
    struct tm *start, // Hora de inicio
    struct tm *end,   // Hora de termino
    double error,     // Porcentaje de error
    double elapsed,   // Tiempo que paso
    int N             // Dimension de la matriz
)
{
    FILE *file;
    char file_name[64];
    char output[50];

    sprintf(file_name, "serial-c-%d-%d-%d-%d-%d.txt", N, end->tm_mday, end->tm_hour, end->tm_min, end->tm_sec);

    file = fopen(file_name, "w+");

    strftime(output, sizeof(output), "%c", start);
    fprintf(file, "Hora de inicio\n%s\n", output);

    strftime(output, sizeof(output), "%c", end);
    fprintf(file, "Hora de termino\n%s\n", output);

    fprintf(file, "Tiempo de ejecucion\n%.15lf\n", elapsed);

    fprintf(file, "Error\n%.15le\n", error);

    fclose(file);
}

File: 3-matmult/pointers/matmult.f90
MODULE precision
  INTEGER ,PARAMETER:: long=SELECTED_REAL_KIND(15,307) !(6,37) for real(float), (15,307) for double
END MODULE precision

MODULE parameters
  USE precision
  IMPLICIT NONE
  REAL(long), PARAMETER :: a = 1.0E-10_long         ! Valor constante para llenar la matriz
END MODULE parameters

module time_ftcs
    contains
      function timestamp() result(str)
        implicit none
        character(len=20)                     :: str
        integer                               :: values(8)
        character(len=4)                      :: year
        character(len=2)                      :: month
        character(len=2)                      :: day, hour, minute, second
        character(len=5)                      :: zone
    
        ! Get current time
        call date_and_time(VALUES=values, ZONE=zone)  
        write(year,'(i4.4)')    values(1)
        write(month,'(i2.2)')   values(2)
        write(day,'(i2.2)')     values(3)
        write(hour,'(i2.2)')    values(5)
        write(minute,'(i2.2)')  values(6)
        write(second,'(i2.2)')  values(7)
    
        str = day//'-'//hour//'-'//minute//'-'//second
      end function timestamp
end module

program matmult
    use precision
    use parameters
    use time_ftcs
    implicit none
    
    INTERFACE 
        subroutine FillMatrix(matrixA, matrixB, N)
            use precision
            REAL(long), INTENT(OUT) :: matrixA(:)
            REAL(long), INTENT(OUT) :: matrixB(:)
            INTEGER, INTENT(IN) :: N
        end subroutine FillMatrix

        subroutine Multiply(MatrixA, MatrixB, MatrixC, N)
            use precision
            REAL(long), INTENT(IN) :: MatrixA(:)
            REAL(long), INTENT(IN) :: MatrixB(:)
            REAL(long), INTENT(OUT) :: MatrixC(:)
            INTEGER, INTENT(IN) :: N
        end subroutine Multiply

        subroutine AddMatrix(matrix, N, result)
            use precision
            REAL(long), INTENT(IN) :: matrix(:)
            INTEGER, INTENT(IN) :: N
            REAL(long), INTENT(OUT) :: result
        end subroutine AddMatrix
    END INTERFACE

    INTEGER :: N                                        ! Dimension de la matriz
    REAL(long), POINTER, DIMENSION(:) :: matrixA        ! Primera matriz
    REAL(long), POINTER, DIMENSION(:) :: matrixB        ! Segunda matriz
    REAL(long), POINTER, DIMENSION(:) :: matrixC        ! Matriz resultado
    REAL(long) :: result                                ! Resultado de la suma de la matriz resultado
    REAL(long) :: estimation                            ! Estimacion del calculo
    REAL(long) :: error                                 ! Error encontrado
    REAL(long) :: elapsed                               ! Tiempo que tomo ejecutarse el programa
    REAL(long) :: start_clock                           ! Tiempo de inicio
    REAL(long) :: stop_clock                            ! Tiempo de termino
    CHARACTER(len=38) :: start                          ! Hora de inicio
    CHARACTER(len=38) :: end                            ! Hora de termino

    ! Establece la hora de inicio
    start = timestamp()
    call CPU_TIME(start_clock)

    ! Asigna la dimension de la matriz
    call OpenFile(N)

    ! Reserva la memoria para las tres matrices
    allocate(matrixA(N * N))
    allocate(matrixB(N * N))
    allocate(matrixC(N * N))

    ! Llena la matriz A y B con el valor constante
    call FillMatrix(matrixA,matrixB, N)
    
    ! Multiplica las matrices A y B guardando el valor en la matriz C
    call Multiply(MatrixA, MatrixB, MatrixC, N)
    
    ! Calcula la suma de los valores de la matriz C
    call AddMatrix(MatrixC, N, result)

    ! Calculo estimado con la formula a^2*N^3.
    estimation = N ** 3_long * a ** 2_long

    ! Calcula el % de error
    error = DABS(result - estimation) / estimation * 100.0_long

    ! Libera la memoria de las tres matrices
    deallocate(matrixA)
    deallocate(matrixB)
    deallocate(matrixC)

    ! Establece la hora en que termino de calcular
    end = timestamp()
    call CPU_TIME(stop_clock)

    ! Calcula el tiempo que tomo ejecutar el programa
    elapsed =  stop_clock - start_clock

    ! Guarda el resultado en un archivo
    call SaveFile(start, end, error, elapsed, N)

end program matmult

! Llena las dos matrices con el valor constante
subroutine FillMatrix(matrixA, matrixB, N)
    use precision
    use parameters
    IMPLICIT NONE
    REAL(long), INTENT(OUT) :: matrixA(:)   ! Primera matriz
    REAL(long), INTENT(OUT) :: matrixB(:)   ! Segunda matriz
    INTEGER, INTENT(IN) :: N                ! Dimension de la matriz
    INTEGER :: i                            ! Indice el renglon
    INTEGER :: j                            ! Indice de la columna

    do i = 1, N, 1
        do j = 1, N, 1
            matrixA((i - 1) * N + j)  = a
            matrixB((i - 1) * N + j)  = a
        end do
    end do

end subroutine FillMatrix

! Multiplica las dos matrices y almacena el resultado en la matriz de resultado
subroutine Multiply(MatrixA, MatrixB, MatrixC, N)
    use precision
    IMPLICIT NONE
    REAL(long), INTENT(IN) :: MatrixA(:)    ! Primera matriz
    REAL(long), INTENT(IN) :: MatrixB(:)    ! Segunda matriz
    REAL(long), INTENT(OUT) :: MatrixC(:)   ! Matriz resultado
    INTEGER, INTENT(IN) :: N                ! Dimension de la matriz
    INTEGER :: i                            ! Indice del renglon
    INTEGER :: j                            ! Indice de la columna
    INTEGER :: k                            ! Indice de la multiplicacion
    REAL(long) :: result                    ! Resultado de la multiplicacion

    do i = 1, N, 1
        do j = 1, N, 1
            result = 0.0
            do k = 1, N, 1
                result = result + matrixA((i - 1) * N + k) * matrixB((k - 1) * N + j)
            end do
            matrixC((i - 1) * N + j) = result
        end do
    end do

end subroutine Multiply

! Suma todos los elementos de una matriz y regresa el resultado
subroutine AddMatrix(matrix, N, result)
    use precision
    IMPLICIT NONE
    REAL(long), INTENT(IN) :: matrix(:)     ! Matriz resultado
    INTEGER, INTENT(IN) :: N                ! Dimension de la matriz
    REAL(long), INTENT(OUT) :: result       ! Resultado de la suma
    INTEGER :: i                            ! Indice del renglon
    INTEGER :: j                            ! Indice de la columna

    result = 0.0_long

    do i = 1, N, 1
        do j = 1, N, 1
            result = result + matrix((i - 1) * N + j)
        end do
    end do

end subroutine AddMatrix

! Abre un archivo con la dimension de la matriz
subroutine OpenFile(N)
    implicit none
    INTEGER, INTENT(OUT) :: N               ! Dimension de la matriz

    OPEN(UNIT=1,file='parameters.dat',ACTION="READ",STATUS='OLD')
    READ(1,*) N
    CLOSE(1)

end subroutine OpenFile

! Crea un archivo de salida con la hora de inicio, de termino y el tiempo que tomo correr el programa
! Asi como el porcentaje de error
subroutine SaveFile(start, end, error, elapsed, N)
    use precision
    CHARACTER(len=38), INTENT(IN) :: start
    CHARACTER(len=38), INTENT(IN) :: end
    REAL(long), INTENT(IN) :: error
    REAL(long), INTENT(IN) :: elapsed
    INTEGER, INTENT(IN) :: N
    CHARACTER(len = 54) :: file_name
    CHARACTER(len = 4) :: Nstring

    write(Nstring, "(I4)") N 

    file_name = 'serial-f90-'//Nstring//'-'//trim(end)//'.txt'
    
    open (unit=10,file=file_name)
    write(10,*) 'Hora de inicio'
    write(10,*) start
    write(10,*) 'Hora de termino'
    write(10,*) end
    write(10,*) 'Tiempo de ejecucion'
    write(10,*) elapsed
    write(10,*) 'Error'
    write(10,*) error
    close(10)

end subroutine SaveFile